examples/natmod/stm32wb55_transparent_vcp/_stm32wb55_transparent.c

```diff
@@ -0,0 +1,639 @@
+#define LHCI_OPCODE_C1_READ_REG ((LHCI_OGF << 10) + LHCI_OCF_C1_READ_REG)
+typedef PACKED_STRUCT {
+    LHCI_Busw_t busw;
+    uint32_t add;
+}
+LHCI_C1_Read_Register_cmd_t;
+
+typedef PACKED_STRUCT {
+    uint8_t status;
+    uint32_t val;
+}
+LHCI_C1_Read_Register_ccrp_t;
+
+#define LHCI_OCF_C1_DEVICE_INF (LHCI_OCF_BASE + 2)
+#define LHCI_OPCODE_C1_DEVICE_INF ((LHCI_OGF << 10) + LHCI_OCF_C1_DEVICE_INF)
+typedef PACKED_STRUCT {
+    uint8_t status;
+    uint16_t rev_id;      /* from DBGMCU_ICODE */
+    uint16_t dev_code_id; /* from DBGMCU_ICODE */
+    uint8_t package_type; /* from package data register */
+    uint8_t device_type_id; /* from FLASH UID64 */
+    uint32_t st_company_id; /* from FLASH UID64 */
+    uint32_t uid64;       /* from FLASH UID64 */
+    uint32_t uid96_0;     /* from Unique device ID register */
+    uint32_t uid96_1;     /* from Unique device ID register */
+    uint32_t uid96_2;     /* from Unique device ID register */
+    MB_SafeBootInfoTable_t SafeBootInf;
+    MB_FusInfoTable_t FusInf;
+    MB_WirelessFwInfoTable_t WirelessFwInf;
+    uint32_t AppFwInf;
+}
+LHCI_C1_Device_Information_ccrp_t;
+
+#define TL_BLECMD_PKT_TYPE (0x01)
+#define TL_ACL_DATA_PKT_TYPE (0x02)
+#define TL_BLEEVT_PKT_TYPE (0x04)
+#define TL_OTCMD_PKT_TYPE (0x08)
+#define TL_OTRSP_PKT_TYPE (0x09)
+#define TL_CLICMD_PKT_TYPE (0x0A)
+#define TL_OTNOT_PKT_TYPE (0x0C)
+#define TL_OTACK_PKT_TYPE (0x0D)
+#define TL_CLINOT_PKT_TYPE (0x0E)
+#define TL_CLIACK_PKT_TYPE (0x0F)
+#define TL_SYSCMD_PKT_TYPE (0x10)
+#define TL_SYSRSP_PKT_TYPE (0x11)
+#define TL_SYSEVT_PKT_TYPE (0x12)
+#define TL_CLIRESP_PKT_TYPE (0x15)
+#define TL_M0CMD_PKT_TYPE (0x16)
+#define TL_LOCCMD_PKT_TYPE (0x20)
+#define TL_LOCRSP_PKT_TYPE (0x21)
+#define TL_TRACES_APP_PKT_TYPE (0x40)
+#define TL_TRACES_WL_PKT_TYPE (0x41)
+
+#define TL_CMD_HDR_SIZE (4)
+#define TL_EVT_HDR_SIZE (3)
+#define TL_EVT_CS_PAYLOAD_SIZE (4)
+
+#define TL_BLEEVT_CC_OPCODE (0x0E)
+#define TL_BLEEVT_CS_OPCODE (0x0F)
+#define TL_BLEEVT_VS_OPCODE (0xFF)
+
+
+typedef PACKED_STRUCT {
+    uint16_t cmdcode;
+    uint8_t plen;
+    uint8_t payload[255];
+}
+TL_Cmd_t;
+
+typedef PACKED_STRUCT {
+    uint8_t type;
+    TL_Cmd_t cmd;
+}
+TL_CmdPacket_t;
+
+typedef PACKED_STRUCT {
+    uint8_t numcmd;
+    uint16_t cmdcode;
+    uint8_t payload[1];
+}
+TL_CcEvt_t;
+
+typedef PACKED_STRUCT {
+    uint8_t evtcode;
+    uint8_t plen;
+    uint8_t payload[1];
+}
+TL_Evt_t;
+
+typedef PACKED_STRUCT {
+    uint8_t type;
+    TL_Evt_t evt;
+}
+TL_EvtSerial_t;
+
+// Callback python function, can be used to blink LED etc on each comm
+mp_obj_t callback_o;
+
+void mpy_cb(bool on) {
+    if (callback_o != mp_const_none) {
+        mp_call_function_n_kw(callback_o, 1, 0, (mp_obj_t[]) {(on) ? mp_const_true : mp_const_false});
+    }
+}
+
+// Stream I/O
+
+mp_obj_t stream_in_o;
+const mp_stream_p_t *stream_in_p;
+mp_obj_t stream_out_o;
+const mp_stream_p_t *stream_out_p;
+
+mp_uint_t mpstream_write(const void *buf, size_t len) {
```

Comment:
Please call it `mp_stream_write`, also same for other 2 functions below.

Also I think it might be cleaner to pass in as the first argument the stream object, instead of having global variables.

See eg `examples/natmod/btree/btree_c.c:mp_stream_posix_write`

Reply:
Ah yes, definitely. This was originally used in the callback in `parse_hci_info_t parse` directly in `rfcore.c` when I was using that directly. I forgot to unwind all of this once I added the rfcore_ble_hci micropython interface, now that it's always called directly it'll be easy enough to pass the stream through.

Reply:
Yep this is fixed now making it much cleaner overall.